pkg/api/frontend_logging.go

```diff
@@ -28,14 +66,121 @@ type frontendSentryEvent struct {
 	Exception *frontendSentryException `json:"exception,omitempty"`
 }
 
+func guessSourceMapLocation(sourceUrl string) (*sourceMapLocation, error) {
+	u, err := url.Parse(sourceUrl)
+	if err != nil {
+		return nil, err
+	}
+	if strings.HasPrefix(u.Path, "/public/build/") {
+		return &sourceMapLocation{
+			dir:      setting.StaticRootPath,
+			path:     path.Join("build", u.Path[len("/public/build/"):]) + ".map",
+			pluginId: "",
+		}, nil
+	} else if strings.HasPrefix(u.Path, "/public/plugins/") {
+		for _, route := range plugins.StaticRoutes {
+			pluginPrefix := path.Join("/public/plugins/", route.PluginId)
+			if strings.HasPrefix(u.Path, pluginPrefix) {
+				return &sourceMapLocation{
+					dir:      route.Directory,
+					path:     u.Path[len(pluginPrefix):] + ".map",
+					pluginId: route.PluginId,
+				}, nil
+			}
+		}
+	}
+	return nil, nil
+}
+
+func getSourceMap(sourceUrl string) (*sourceMap, error) {
+	sourceMapCache.Lock()
+	defer sourceMapCache.Unlock()
+
+	if smap, ok := sourceMapCache.cache[sourceUrl]; ok {
+		return smap, nil
+	}
+	sourceMapLocation, err := guessSourceMapLocation(sourceUrl)
+	if err != nil {
+		return nil, err
+	}
+	if sourceMapLocation != nil {
+		dir := http.Dir(sourceMapLocation.dir)
+		f, err := dir.Open(sourceMapLocation.path)
+		if err != nil {
+			if os.IsNotExist(err) {
+				sourceMapCache.cache[sourceUrl] = nil
+				return nil, nil
+			}
+			return nil, err
+		}
+		defer func() {
+			if err := f.Close(); err != nil {
+				frontendLogger.Error("Failed to close source map file.", "err", err)
```

Comment:
We generally don't punctuate log messages:
```suggestion
				frontendLogger.Error("Failed to close source map file", "err", err)
```

Reply:
updated!